src/py2spack/loading.py

```diff
@@ -20,166 +21,257 @@
             timeout=10,
         )
-        if r.status_code != HTTP_STATUS_SUCCESS:
-            if r.status_code == HTTP_STATUS_NOT_FOUND:
+        if r.status_code != utils.HTTP_STATUS_SUCCESS:
+            if r.status_code == utils.HTTP_STATUS_NOT_FOUND:
                 msg = f"Package {name} not found on PyPI (status code 404)"
-                return APIError(msg)
+                return PyProjectProviderQueryError(msg)
 
             msg = (
                 f"Error when querying JSON API (status code {r.status_code})."
                 f" Response: {r.text}"
             )
-            return APIError(msg)
+            return PyProjectProviderQueryError(msg)
+
+        data: dict[Any, Any] = r.json()
+        return data
+
+    @functools.cache  # noqa: B019
+    def get_versions(self, name: str) -> list[vn.Version] | PyProjectProviderQueryError:
+        """Get usable versions for package.
+
+        Returns an error if no versions are found.
+        In addition to the caching of the `_get` method, we also cache all calls
+        to `get_versions`, because the versions are needed frequently during the
+        conversion process for dependencies, and the size of the data is small.
+        """
+        data = self._get(name)
+        if isinstance(data, PyProjectProviderQueryError):
+            return data
 
-        data: dict = r.json()
         versions = data["versions"]
 
         # parse and sort versions
-        parsed_versions = sorted({vv for v in versions if (vv := _acceptable_version(v))})
+        result: list[vn.Version] | PyProjectProviderQueryError = sorted(
+            {vv for v in versions if (vv := _parse_packaging_version(v))}
+        )
 
-        self.version_cache[name] = parsed_versions
+        if not result:
+            result = PyProjectProviderQueryError("No valid versions found")
 
-        return data
+        return result
 
-    def get_versions(self, name: str) -> list[vn.Version]:
-        """Get acceptable versions for package (from cache if possible)."""
-        if name not in self.version_cache:
-            data = self._get(name)
+    def get_pyproject(
+        self, name: str, version: vn.Version
+    ) -> dict[Any, Any] | PyProjectProviderQueryError:
+        """Download and extract the pyproject.toml for the specified package version."""
+        all_metadata = self._get_distribution_metadata(name)
 
-            if isinstance(data, APIError):
-                self.version_cache[name] = []
+        if isinstance(all_metadata, PyProjectProviderQueryError):
+            return all_metadata
 
-        return self.version_cache[name]
+        metadata = all_metadata[version]
+        # python sdist archives contain a top level directory, e.g. "black-24.4.2/"
+        directory_str = f"{name}-{version}"
 
-    def get_files(self, name: str, last_n_versions: int = -1) -> list[dict[Any, Any]] | APIError:
+        # for type checker, we know these values are going to be strings
```

Comment:
I am not sure if this is the right way to do it, but I added these asserts because we know that the values here are going to be strings (dict is created by us earlier) and otherwise the type checker complains